### PR TITLE
registration: add function to re-request email token

### PR DESCRIPTION
**Pull request by justjanne**
Type: Enhancement
Related: https://github.com/vector-im/element-web/issues/21984
Related: https://github.com/matrix-org/matrix-react-sdk/pull/8554

---
Here's what your changelog entry will look like:

## ✨ Features
 * registration: add function to re-request email token ([\#2357](https://github.com/matrix-org/matrix-js-sdk/pull/2357)). Contributed by @justjanne.

### CODE REVIEW

File: src/interactive-auth.ts [hunk 1]
```diff
@@ -408,6 +410,34 @@ export class InteractiveAuth {
         this.emailSid = sid;
     }
 
+    /**
+     * Requests a new email token and sets the email sid for the validation session
+     */
+    public requestEmailToken = async () => {
```
Comment (t3chguy) [inline src/interactive-auth.ts:416]: out of interest is there any reason this is an arrow func?

**Review by t3chguy — COMMENTED** [review]
Looks sane otherwise, but as pointed out by Sonar, does not hit the required mark of 80% coverage on new code.

File: src/interactive-auth.ts (hunk 1 above)
Comment (justjanne) [reply src/interactive-auth.ts:416]: We need to use `this` in the function, and while I personally would prefer to use a regular function and binding it in the constructor, across the codebase I’ve found arrow functions to be used more commonly to accomplish the same.

Comment (t3chguy) [reply src/interactive-auth.ts:416]: Sure but the function is always called with the correct `this` context already (L493) so would work in either way. Its not being copied to lose its bound context anywhere yet

Comment (t3chguy) [reply src/interactive-auth.ts:416]: The SDK as a whole makes the assumption that the caller will always call with the correct `this` context, we tend to use arrow funcs for when we want to use it as a local handler. Arrow functions here are not wrong by any means, was just curious.

**Review by t3chguy — APPROVED** [review]
🥳 Thanks for being the first to brave the better tested new world order!

File: src/interactive-auth.ts (hunk 1 above)
Comment (justjanne) [reply src/interactive-auth.ts:416]: Oh, that’s an interesting assumption! Good to know, I’ve never seen a codebase that assumed this before.

**Comment by justjanne** [comment]
Sadly I couldn’t migrate the tests to typescript cause they do some type trickery :(

File: src/interactive-auth.ts (hunk 1 above)
Comment (t3chguy) [reply src/interactive-auth.ts:416]: It is very much a bad assumption to have, but one we make anyway. If you feel strongly about it we could make a rule that all public methods have to be arrow functions to remedy this

**Pull request merged**